### DISCUSSION TITLE
JWT Verification Fails Error on Istio

### DISCUSSION THREAD

**Discussion by Cascaida** [istio/istio#51098] [Q&A]
I am trying to setup JWT authentication using Istio. I've configured RequestAuthentication resource for enabling JWT authentication.

Below I am sharing the YAML file content of the RequestAuthentication resource that I've created.

`apiVersion: security.istio.io/v1beta1
kind: RequestAuthentication
metadata:
  name: <RequestAuthenticationResourceName>
  namespace: <Namespace>
spec:
  selector:
    matchLabels:
      app: <ApplicationName>
  jwtRules:
    - forwardOriginalToken: true
      issuer: <IssuerURL>
      jwks: |
        {"keys": [{"n": "<SecretAppliedHere>", "e": "AQAB", "kty": "RSA"}]}`

We just received the JWT secret and changed it into Base64 using online Base64 converter websites. The YAML is applied successfully this way.

Then we logged in to a testing pod and tried to send a CURL request to the internal service URL of the pod that we used in "matchLabels" section in RequestAuthentication resource with a valid token. ( we have verified that token is valid with sending another request to another endpoint that accepts the token )

Below I am sharing the CURL command that we used for testing JWT authentication with Istio.

`curl --location '
[http://<internal-service-url-and-path-applied-here>
--header 'accept: application/json, text/plain, */*' --header 'authorization: Bearer <token-applied-here> --header 'Cookie: <cookie-applied-here>'`

After performing this CURL operation, we are getting "JWT Verification Fails Error" in the output. 

I'd be glad if someone can point the correct direction or help with this issue.

**Comment by amitborole88** [comment]
https://istio.io/latest/docs/reference/config/security/request_authentication/

Your yaml  has jwks. It should be jwksUri.

apiVersion: security.istio.io/v1beta1 kind: RequestAuthentication metadata: name: <RequestAuthenticationResourceName> namespace: <Namespace> spec: selector: matchLabels: app: <ApplicationName> jwtRules: - forwardOriginalToken: true issuer: <IssuerURL> jwksUri: <<  issuer URL  key URI >>

**Reply by Cascaida** [reply]
thanks for your answer @amitborole88 , is it mandatory to enter jwksUri? Because instead of providing the secret from a URI, we would like to submit the secret directly using jwks.

**Reply by amitborole88** [reply]
No. Its not mandatory. However, when tested with jwksUri it has worked for me.

To troubleshoot further, you can enable the debug logs in your istio proxy and check in it.

Sample command: istioctl proxy-config log <PODNAME> -n <namespace> --level=debug
